### ISSUE TITLE
Easy fix for making it usable with a NetworkStream

### ISSUE THREAD

**Issue by jafaber** [MiloszKrajewski/lz4net#2]
`LZ4Stream` can currently not really be wrapped around a `System.IO.NetworkStream` because the receiving side will keep blocking trying to get the next chunk. The blocking is only stopped after the buffer is full or the stream is closed. Neither happens often with an interactive TCP session.

One simple solution I just tested is by letting the `Read()` return after successfully reading one chunk and NOT letting it wait for another chunk.

For that I added a public property `bool ReadPerChunk` and added this line to the `Read()` method:

`if (ReadPerChunk) break;`  (as the last line of `if (chunk > 0) {  }` )

I guess I could have made a pull request, but this seems quite hacky and maybe you'd prefer a different way (or something more compatible with the other language implementations).

You could argue that it would be more efficient to try to read as many chunks as possible before giving up, but you'd have to check the `NetworkStream.DataAvailable` property, which is not available in the `Stream` interface (so it would require checking types and casting etc., which would probably kill your performance gain).

I have considered automatically detecting a `NetworkStream` or making `LZ4Stream` generic for the Stream type, but I don't think that will suffice as there might be for example a `BufferedStream` wrapped around the `NetworkStream`.

**Comment by jafaber** [comment]
I ended up adding an extra argument to the constructor (`readPerChunk`) instead of a public property. Beneath is my test Console application:

```c#
using System;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using LZ4;

namespace ConsoleApplication6 {
  class Program {
    private static TcpClient tcpclient;
    private static TcpListener tcplistener;
    private static NetworkStream mNetworkStream;
    private static byte[] mDecompressedBuffer = new byte[2048];

    static void Main(string[] args) {
      bool lIsClient = args.Count() > 0;

      if (lIsClient) {
        tcpclient = new TcpClient();
        tcpclient.NoDelay = true;
        Console.WriteLine("Connecting...");
        tcpclient.Connect(args[0], 5000);
      } else {
        tcplistener = new TcpListener(new System.Net.IPAddress(0), 5000);
        tcplistener.Start();
        Console.WriteLine("Listening...");
        tcpclient = tcplistener.AcceptTcpClient();
        tcpclient.NoDelay = true;      Console.WriteLine("Done, socket statistics: {0} bytes sent, {1} bytes received", cw.CompressedByteCount, cr.CompressedByteCount);

      }

      Console.WriteLine("Connected");

      mNetworkStream = tcpclient.GetStream();
      Console.WriteLine("mNetworkStream: {0} {1}", mNetworkStream.CanRead, mNetworkStream.CanWrite);

      //BufferedStream br = new BufferedStream(mNetworkStream);
      //BufferedStream bw = new BufferedStream(mNetworkStream);

      LZ4.LZ4Stream cr = new LZ4Stream(mNetworkStream, LZ4.LZ4StreamMode.Decompress
                                        , readPerChunk: true);
      LZ4.LZ4Stream cw = new LZ4Stream(mNetworkStream, LZ4.LZ4StreamMode.Compress);

      Console.WriteLine("Sending...");

      int lFlushes = 0;
      byte[] buffer;
      for (int i = 0; i < 2; i++) {
        buffer = ASCIIEncoding.ASCII.GetBytes("Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello Hello !");
        cw.Write(buffer, 0, buffer.Length);
        cw.Flush(); lFlushes++;
      }

      Console.WriteLine("Sent {0} chunks", lFlushes);
      //if (lIsClient) { Console.WriteLine("Press key to continue..."); Console.ReadKey(); }

      Console.WriteLine("Receiving...");

      for (int i = 0; i < lFlushes; i++) {
        int l = cr.Read(mDecompressedBuffer, 0, mDecompressedBuffer.Length);

        Console.WriteLine("Received chunk of {0} bytes", l);
        Console.WriteLine(ASCIIEncoding.ASCII.GetString(mDecompressedBuffer, 0, l));
      }

      Console.WriteLine("Press key to continue..."); Console.ReadKey();
    }
  }
}
```

**Comment by MiloszKrajewski** [comment]
Hi,

I'm not sure if I understand what problem you are facing. I didn't test it (I will, I just don't have time this week) but I believe it reads chuck by chuck from stream (as it needs whole chunk to decompress) and when whole chunk i available it decompresses it and provides data to the reader. As long there is enough data in buffer it won't block again.

Do you want chunk size to be dynamic? If one end writes 1 byte then other end should immediately see it? It would impact compress ratio badly. In such case I would use Wrap / Unwrap to encode / decode packets and use NetworkStream directly.

Did I get it right?

**Comment by jafaber** [comment]
No, not dynamic chunks. Chunk by chunk is fine. I do a Flush() regularly on the sending side to create the chunks. The bug is actually here (quoting your description) :

> ... and when whole chunk i available it decompresses it and provides data to the reader.

It actually doesn't provide the data back to the reader, instead it immediately starts reading the next chunk (and thus blocking waiting for the next chunk). That's why I'm suggesting inserting a `break` after one chunk is finished.

But please don't hurry. I fixed it for my case already, so I'm good. Once you have some time and look at the code, I'm sure you'll see what I mean.

BTW, I also tried another C# wrapper and it turned out to have the exact same bug, just in C++ code this time. I'm not sure how far upstream this bug goes.  I reported it there too: https://github.com/IonKiwi/lz4.net/issues/2

Thanks!

**Comment by MiloszKrajewski** [comment]
I took a glimpse (still didn't test it), but correct if I'm wrong:

You would like Read on LZ4Stream to return with any number of bytes of bytes as long as it is more than 0. Current implementation tries to read exactly as much bytes as requested and blocks until you have that many bytes while the contract on read should be Read UP TO that many bytes?

Am I right? Is it what you are raising? If so... I agree - this would be correct behavior :-) I'm just worried about those people who rely on current implementation.

**Comment by jafaber** [comment]
Yes. Exactly.

This is actually how System.NetworkStream works as well. If you do Read(,, 100) you get UP TO 100 bytes, but if there are only 10 bytes available and the socket simply doesn't have any more, it will return with 10 bytes. Only when you do another Read(,,100) and there are no bytes available at all, then it will block and wait for bytes.

For reading files or HTTP streams there's not much difference, but in interactive TCP sessions (or any other stream where the sender is sending piece by piece over a longer period of time) you can't have any blocking while there's data in buffers.

To keep it backwards compatible I suggested making my change optional (with a property or constructor argument). But I'll let you (or upstream) decide whether that's necessary. It actually makes sense to do it like NetworkStream does.

**Comment by MiloszKrajewski** [comment]
FYI: has been addressed but neither tested or relased

**Comment by MiloszKrajewski** [comment]
Fixed in 1.0.10.93, see LZ4StreamFlags.InteractiveRead

NOTE: In my opinion what you are describing should be default behaviour, but I was a little bit worried about code relying on old bevaviour (reading as many bytes as requested) that I made it optional; Mt heart tells me DEFAULT but my pragmatism suggests OPTIONAL
